test/src/test/java/org/corfudb/runtime/object/transactions/SequencerServerCacheTest.java

```diff
@@ -68,36 +64,152 @@ public void testSequencerCacheTrim() {
+            if (oldAddress < cache.firstAddress()) {
+                continue;
+            }
+            ConflictTxStream key = new ConflictTxStream(oldKey.getStreamId(), oldKey.getConflictParam(), 0);
+            log.debug("address " + cache.get(key) + " expected " + oldAddress);
+            assertThat(cache.get(key) == oldAddress);
+        }
+    }
+
+    @Test
+    /*
+        Test the evication the firstAddress while the cache is full, by generating address out of order
+     */
+    public void testSequencerCacheEvict1() {
+        SequencerServerCache cache = new SequencerServerCache(cacheSize, Address.NOT_FOUND);
+        long address = 0;
+        HashMap<ConflictTxStream, Long> recordMap = new HashMap<>();
+
+        // put entries to the cache with duplicate address not in order
+        while (cache.size() < cacheSize) {
+            address = 0;
+            for (int i = 0; i < cacheSize / entryPerAddress; i++) {
+                generateData(recordMap, cache, address++, false);
+            }
+        }
+
+        assertThat(cache.size() == cacheSize);
+        verifyData(recordMap, cache);
+
+        // Each put should evict all streams with the same address
+        for (int i = 0; i < iterations; i++, address++) {
+            generateData(recordMap, cache, address, true);
+        }
+
+        verifyData(recordMap, cache);
+
+        cache.invalidateUpTo(address - 1);
+        assertThat(cache.size() == 1);
+        cache.invalidateUpTo(address);
+        assertThat(cache.size() == 0);
+    }
+
+    @Test
+    /*
+        Test the evication the firstAddress while the cache is full, by generating address in order
+     */
+    public void testSequencerCacheEvict2() {
+        SequencerServerCache cache = new SequencerServerCache(cacheSize, Address.NOT_FOUND);
+        long address = 0;
+        HashMap<ConflictTxStream, Long> recordMap = new HashMap<>();
+
+        // put entries to the cache, make it full, some entries have the same address
+        while (cache.size() < cacheSize) {
+            for (int j = 0; j < entryPerAddress; j++) {
+                generateData(recordMap, cache, address++, false);
+            }
+        }
+
+        verifyData(recordMap, cache);
+
+        assertThat(cache.size() == cacheSize);
+        // Each put should evict all streams with the same address
+        for (int i = 0; i < iterations; i++, address++) {
+            generateData(recordMap, cache, address, true);
+        }
+
+        verifyData(recordMap, cache);
+        log.info("cacheSize {} cacheByteSize {} cacheEntriesBytes {} ", cache.size(), cache.byteSize(), cache.byteSize());
+        long entrySize = cache.byteSize() / cache.size();
+
+        cache.invalidateUpTo(address - numRemains);
+        assertThat(cache.size() == numRemains);
+
+        // this assume that the all conflickstreams has the same size of the parameters.
+        assertThat(entrySize == cache.byteSize() / cache.size());
```

Comment:
FYI, I did experiment about `assertThat` and looks `assertThat` does not assert its parameter is true.